src/ducc0/infra/threading.cc

```diff
@@ -367,24 +372,29 @@ class thread_pool
       }
   };
 
-inline thread_pool &get_pool()
+// return a pointer to a singleton ducc_thread_pool, which is always available
+inline std::shared_ptr<ducc_thread_pool> get_master_pool()
   {
-  static thread_pool pool;
+  static auto master_pool=std::make_shared<ducc_thread_pool>();
 #if __has_include(<pthread.h>)
   static std::once_flag f;
   call_once(f,
     []{
     pthread_atfork(
-      +[]{ get_pool().shutdown(); },  // prepare
-      +[]{ get_pool().restart(); },   // parent
-      +[]{ get_pool().restart(); }    // child
+      +[]{ get_master_pool()->shutdown(); },  // prepare
+      +[]{ get_master_pool()->restart(); },   // parent
+      +[]{ get_master_pool()->restart(); }    // child
       );
     });
 #endif
-
-  return pool;
+  return master_pool;
   }
 
+thread_local std::shared_ptr<thread_pool> active_pool = get_master_pool();
```

Comment:
I think I understand why you want to use smart pointers here - because a copy is taken below within `thread_map` and you want to guarantee the pool is still valid.

The shared pointer adds some overhead (although very minimal in this case), but more importantly it forces the caller to share ownership of _its_ threadpool, which the caller may not want to do.  If the caller destroys its threadpool before the task is complete, I'd say that's a bug in _their_ code, not yours.  As long as you make the ownership contract clear, I think a raw pointer should suffice.

Reply:
I don't see how the caller can destroy the thread pool before the completion of the task, because control cannot leave `thread_map` while the task is running, but I'm probably missing something.

Anyway I think I can live with raw pointers here as well. Should we change the `PoolGuard` constructor to

`PoolGuard(thread_pool &pool);`

and then take the address inside the constructor, or simply do a

`PoolGuard(thread_pool *pool);`

The first version feels intuitively better to me; it excludes null pointers at least.

Reply:
> I don't see how the caller can destroy the thread pool before the completion of the task

Right, the only way I could think of is if they did something like
```
PoolGuard(std::make_unique<MyCustomThreadPool>().get());
```
Or if they explicitly delete the pool prior to first use.  Otherwise, I couldn't really think of a safety reason why we would need `shared_ptr`.

> Should we change the PoolGuard constructor to `PoolGuard(thread_pool &pool);`

Yes, I think so.  This forces an lvalue of the pool to exist.  Maybe document `PoolGuard` to explicitly state that it stores a reference to the input, so the pool must remain alive until it is reset.

Maybe the name of `PoolGuard` should be a bit more explicit about what it's used for?  It's not for managing the life of the threadpool itself, but for explicitly setting the threadpool to use within the provided scope.  Something like `ScopedUseThreadPool`?

Reply:
I'm using raw pointers now for the thread pools; I also changed the class name as suggested.